### ISSUE TITLE
Add the type `Module` to emulate the `torch.nn.Module` behavior

### ISSUE THREAD

**Issue by Alphaharrius** [Alphaharrius/PyHilbert#57]
# Descriptions
When we have a `pyhilbert.abstract.Functional` what is represented by highly complex logical operations, sometimes there will be some `Tensor` objects defined within. The `Module` as a subclass of `pyhilbert.abstract.Functional` will provides interface of `.cpu()` and ` .gpu()` and by defaults enabled gradient for all underlying `Tensor` attributes.

# TODOs
- Design `Module` data structure
- Implement device APIs like `.cpu()` and `.gpu()` that is propagative to `Module` attribute under the current `Module`.
- Implement gradient APIs `.requires_grad()` and `.eval()` that is propagative to `Module` attribute under the current `Module`.
- Add the option to use device index to align with torch's definition
